// core/user.js
class User {
	constructor(var1) {
		this._socket = var1.socket;
		this._username = var1.username;
		this._avatar = var1.avatar;
		this._group = null;
		this._permissions = null;
	};

	/**
	 * @param {Websocket} var1
	 */
	setSocket(var1) {
		this._socket = var1;
	};

	/**
	 * @param {string} var1
	 */
	setUsername(var1) {
		this._username = var1;
	};

	/**
	 * @param {integer} var1
	 */
	setAvatar(var1) {
		this._avatar = var1;
	};

	/**
	 * @param {Group} var1
	 */
	setGroup(var1) {
		this._group = var1;
	};

	/**
	 * @param {Array.<string>} var1
	 */
	setPermissions(var1) {
		this._permissions = var1;
	};

	getSocket() {
		return this._socket;
	};

	getUsername() {
		return this._username;
	};

	getGroup() {
		return this._group;
	};

	getPermissions() {
		return this._permissions;
	};

	/**
	 * @param {} var1
	 */
	notify(var1) {
		this._socket.send(var1);
	};

	/**
	 * @param {WebSocket} var1
	 */
	static addUser(var1) {
		let socket = var1;

		socket.on('message', (data) => {
			let parsed_data;

			try {
				parsed_data = JSON.parse(data);
			} catch (e) {
				return;
			}

			let action = parsed_data.action;

			switch(action) {
				case "signup" :
					break;
				
				case "signin" :
					break;

				case "signout" :
					break;

				case "upload" :
					break;
			}
		});

		socket.on('close', () => {
			
		});

		socket.on('message', () => {

		});
	};

	/**
	 * @param {Socket}
	 */
	static getUser(var1) {};
};

module.exports = User;
